feat: add camera panning with constraints - fix #574

diegogliarte opened the pull request:
Uses middle button (wheel button pressed) to pan the camera

![pan](https://github.com/user-attachments/assets/c2e45b1b-8c1f-4adf-b783-a34682b89cda)

pagoru commented:
not a good idea to use the middle button, because it's only available for conventional cursors and the project aims to be more accessible for touch screens.
 It needs to be with the left cursor buuuuuut without compromising the clicking or placing (future) items.

diegogliarte commented:
> not a good idea to use the middle button, because it's only available for conventional cursors and the project aims to be more accessible for touch screens. It needs to be with the left cursor buuuuuut without compromising the clicking or placing (future) items.

I changed it to left-click. To don't move the character while you are moving the camera, I did a little bit of logic on the player mov, check if you think that's the correct approach.

Also extracted the method so we can use it for any component we want.

pagoru commented on app/client/src/shared/utils/camera-panning.utils.ts:
```diff
@@ -0,0 +1,46 @@
+import { DisplayObjectEvent } from "@tu/tulip";
+
+export function allowCameraPanning(component, margin = 0) {
```
prevent the use of `function`, instead use arrow functions and type the params of the function

But, I don't understand why is called this on an independent function instead of inline code, because it's not being used anywhere else and it's not a component neither
We try to avoid creating functions if are not used elsewhere to prevent creating a function with hundred params to do a lot of different things. 

Imagine I have the private rooms, and the public rooms, maybe sharing this function seems a good idea, buut public rooms need an especific param to do their things, and later, private rooms need another param to do another thing, it doesn't make sense to make a lot of exceptions, it was better from the beginning to make indepenent inline code.

diegogliarte replied:
I don't get why public rooms, private rooms, or any other component that is passed through this method would need any "extra params". The logic would be the same for every single component: get the x/y/width/height and allow some sort of camera panning on that element with the mouse, nothing else.

I simply thought that we might have other things like "edit mode", where we would also want to allow moving the camera around, or any other component not related to rooms that you would like to we able to move like that.

pagoru replied:
I mean, it's really not a problem at all, it's more a way of how to structure the project. I'm more of the philosophy of not creating functions if are not really needed. In this case, it can be inline because it's not used elsewhere. If in the future is needed elsewhere, we can move it, but, if this code structure is really not needed anymore elsewhere in the future, it didn't make sense from the beginning to create a function. 

For example, we have this two commands:
https://github.com/openhotel/openhotel/blob/191b00bd3eb435d5e284646e66c0a78187efebcb/app/server/src/modules/system/commands/blacklist.command.ts
https://github.com/openhotel/openhotel/blob/191b00bd3eb435d5e284646e66c0a78187efebcb/app/server/src/modules/system/commands/whitelist.command.ts

At first look, look the same and maybe in your head it's a good idea to make a function to unify everything, but it's not a good idea because are very little variations, and we don't know if we are gonna change something on one or other command, and maintaining one function for both logics it's not a good idea.

We can argue more about this topic on discord if you want, i think it's really interesting! 👯

diegogliarte replied:
Oh, I much prefer avoiding any inline code and extract the methods on as much logic as I can, for both readability and reusability (even though if there's no other case on the program at that moment). I will stick to the project's current approach to maintain consistency!

pagoru commented on app/client/src/shared/utils/camera-panning.utils.ts:
```diff
@@ -0,0 +1,46 @@
+  let dragStart = { x: 0, y: 0 };
+  let containerStart = { x: 0, y: 0 };
+
+  const pixiContainer = component.getDisplayObject({
+    __preventWarning: true,
+  });
```
All the necessary functions used are exposed via tulip , prevent the use of this function at all costs! It's only meant to be used on case of not having other alt

diegogliarte replied:
Done

pagoru commented on app/client/src/shared/utils/camera-panning.utils.ts:
```diff
@@ -0,0 +1,46 @@
+    if (event.button !== 0) return;
+    isDragging = true;
+    dragStart = { x: event.global.x, y: event.global.y };
+    containerStart = { x: pixiContainer.x, y: pixiContainer.y };
```
You can retrieve `pixiContainer.x` with `container.getPosition().x`

diegogliarte replied:
Done

pagoru commented on app/client/src/shared/utils/camera-panning.utils.ts:
```diff
@@ -0,0 +1,46 @@
+    let newX = Math.floor(containerStart.x + deltaX);
+    let newY = Math.floor(containerStart.y + deltaY);
+
+    const minX = -pixiContainer.width / 2 + margin;
```
you can retrieve `pixiContainer.width` with `container.getBounds().width`

diegogliarte replied:
Done

pagoru commented on app/client/src/shared/utils/camera-panning.utils.ts:
```diff
@@ -0,0 +1,46 @@
+    const minY = -pixiContainer.height / 2 + margin;
+    const maxY = pixiContainer.height / 2 - margin;
+
+    pixiContainer.x = Math.max(minX, Math.min(newX, maxX));
+    pixiContainer.y = Math.max(minY, Math.min(newY, maxY));
```
This is really dangerous, because the positions are being calculated on tulip to prevent decimals.
Use `container.setPosition(...)`

diegogliarte replied:
Done

pagoru commented:
And the final thing, it only drags the room when selecting the tiles, I'm not sure if it would be nice from everywhere or not. Maybe controlling the global cursor instead of the containers one (?)

diegogliarte commented:
> And the final thing, it only drags the room when selecting the tiles, I'm not sure if it would be nice from everywhere or not. Maybe controlling the global cursor instead of the containers one (?)

Done in last commit, I don't know how to retrieve the application scale though, which I need to make the drag work.

pagoru commented:
> > And the final thing, it only drags the room when selecting the tiles, I'm not sure if it would be nice from everywhere or not. Maybe controlling the global cursor instead of the containers one (?)
> 
> Done in last commit, I don't know how to retrieve the application scale though, which I need to make the drag work.

nice, with `global.window.getScale();` you can retrieve it

pagoru commented on app/client/src/shared/utils/camera-panning.utils.ts:
```diff
@@ -0,0 +1,42 @@
+    const maxY = height / 2 - margin;
+
+    component.setPositionX(Math.max(minX, Math.min(newX, maxX)));
+    component.setPositionY(Math.max(minY, Math.min(newY, maxY)));
```
You can use `setPosition({ x: 0, y: 0 });` to prevent a double call

diegogliarte replied:
Done

diegogliarte commented:
> > > And the final thing, it only drags the room when selecting the tiles, I'm not sure if it would be nice from everywhere or not. Maybe controlling the global cursor instead of the containers one (?)
> > 
> > 
> > Done in last commit, I don't know how to retrieve the application scale though, which I need to make the drag work.
> 
> nice, with `global.window.getScale();` you can retrieve it

Done

pagoru commented on app/client/src/modules/scenes/private-room/private-room.component.ts:
```diff
@@ -59,6 +60,54 @@ export const privateRoomComponent: ContainerComponent = () => {
+  let dragStart = { x: 0, y: 0 };
+  let containerStart = { x: 0, y: 0 };
+
+  global.events.on(TulipEvent.POINTER_DOWN, (event) => {
```
We need to call this events on `mount` and remove it on `unmount` from `$container`. This is to prevent this events calling when the component is not loaded anymore. You can check this example https://github.com/openhotel/openhotel/blob/191b00bd3eb435d5e284646e66c0a78187efebcb/app/client/src/modules/interfaces/hot-bar.component.ts#L78-L87

pagoru reviewed (commented):
Last fix and all clear to merge! :rocket:
